zkevm-circuits/src/tx_circuit/sign_verify.rs

```diff
@@ -0,0 +1,978 @@
+    pub(crate) msg_hash: secp256k1::Fq,
+}
+
+// Returns (r, s)
+fn sign(
+    randomness: secp256k1::Fq,
+    sk: secp256k1::Fq,
+    msg_hash: secp256k1::Fq,
+) -> (secp256k1::Fq, secp256k1::Fq) {
+    let randomness_inv =
+        Option::<secp256k1::Fq>::from(randomness.invert()).expect("cannot invert randomness");
+    let generator = Secp256k1Affine::generator();
+    let sig_point = generator * randomness;
+    let x = *Option::<Coordinates<_>>::from(sig_point.to_affine().coordinates())
+        .expect("point is the identity")
+        .x();
+
+    let x_repr = &mut Vec::with_capacity(32);
+    x.write(x_repr).expect("cannot write bytes to array");
+
+    let mut x_bytes = [0u8; 64];
+    x_bytes[..32].copy_from_slice(&x_repr[..]);
+
+    let sig_r = secp256k1::Fq::from_bytes_wide(&x_bytes); // get x cordinate (E::Base) on E::Scalar
+    let sig_s = randomness_inv * (msg_hash + sig_r * sk);
+    (sig_r, sig_s)
+}
+
+lazy_static! {
+    static ref SIGN_DATA_DEFAULT: SignData = {
+        let generator = Secp256k1Affine::generator();
+        let sk = secp256k1::Fq::one();
+        let pk = generator * sk;
+        let pk = pk.to_affine();
+        let msg_hash = secp256k1::Fq::one();
+        let randomness = secp256k1::Fq::one();
+        let (sig_r, sig_s) = sign(randomness, sk, msg_hash);
+
+        SignData {
+            signature: (sig_r, sig_s),
+            pk,
+            msg_hash,
+        }
+    };
+}
+
+impl Default for SignData {
+    fn default() -> Self {
+        // Hardcoded valid signature corresponding to a hardcoded private key and
+        // message hash generated from "nothing up my sleeve" values to make the
+        // ECDSA chip pass the constraints, to be use for padding signature
+        // verifications (where the constraints pass, but we don't care about the
+        // message hash and public key).
+        SIGN_DATA_DEFAULT.clone()
+    }
+}
+
+fn pub_key_hash_to_address<F: FieldExt>(pk_hash: &[u8]) -> F {
+    pk_hash[32 - 20..]
+        .iter()
+        .fold(F::zero(), |acc, b| acc * F::from(256) + F::from(*b as u64))
+}
+
+#[cfg(test)]
+mod sign_verify_tests {
+    use super::*;
+    use group::Group;
+    use halo2_proofs::{
+        circuit::SimpleFloorPlanner, dev::MockProver, pairing::bn256::Fr, plonk::Circuit,
+    };
+    use pretty_assertions::assert_eq;
+    use rand::{RngCore, SeedableRng};
+    use rand_xorshift::XorShiftRng;
+
+    #[derive(Clone, Debug)]
+    struct TestCircuitSignVerifyConfig<F: FieldExt> {
+        sign_verify: SignVerifyConfig<F>,
+    }
+
+    impl<F: FieldExt> TestCircuitSignVerifyConfig<F> {
+        pub fn new(meta: &mut ConstraintSystem<F>) -> Self {
+            let power_of_randomness = {
+                let columns = [(); POW_RAND_SIZE].map(|_| meta.instance_column());
+                let mut power_of_randomness = None;
+
+                meta.create_gate("power of randomness", |meta| {
+                    power_of_randomness =
```

Comment:
It seems that this code needs `create_gate` to evaluate the contrains (`|meta|{}`) in the same function, something that seems not guaranteed by the function defintion.

Reply:
Ah, this is the same case as https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/484#discussion_r875662239
This should disappear in the final version once we have a way to get randomness from the transcript.  We do this meanwhile.